pkg/controller/prowjob/prowjob-controller.go

```diff
@@ -0,0 +1,149 @@
+type ProwJobController struct {
+	controller.Controller
+	prowJobLister lister.ProwJobLister
+	// ctx is the passed-down global context. It's used and passed
+	// down to all API client calls as to make sure all in-flight calls get
+	// cancelled if the main context is
+	ctx context.Context
+	// controllerCancel is the components global cancelFunc. It's used to
+	// cancel the global context, stop the leader election and subsequently
+	// initiate a shut down of all control loops
+	controllerCancel context.CancelFunc
+	// Recorder is an event recorder for recording Event resources to the
+	// Kubernetes API.
+	recorder record.EventRecorder
+}
+
+// NewProwJobController returns a new ProwJob controller
+func NewProwJobController(
+	controllerContext context.Context,
+	controllerCancel context.CancelFunc,
+	kubeClientset kubernetes.Interface,
+	prowJobInformer informer.ProwJobInformer,
+	prowJobNamespace string) *controller.Controller {
+
+	utilruntime.Must(prowscheme.AddToScheme(scheme.Scheme))
+
+	// Create event broadcaster
+	klog.V(4).Info("Creating event broadcaster")
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartStructuredLogging(0)
+	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: kubeClientset.CoreV1().Events(prowJobNamespace)})
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: prowJobControllerAgentName})
+
+	prowJobController := &ProwJobController{
+		prowJobLister:    prowJobInformer.Lister(),
+		ctx:              controllerContext,
+		controllerCancel: controllerCancel,
+		recorder:         recorder,
+	}
+
+	c := controller.NewController(
+		[]cache.InformerSynced{prowJobInformer.Informer().HasSynced},
+		prowJobController,
+		prowJobControllerAgentName,
+		reflect.TypeOf(&v1.ProwJob{}),
+	)
+
+	prowJobFilter := func(obj interface{}) bool {
+		if prowJob, ok := obj.(*v1.ProwJob); ok {
+			if _, ok := prowJob.Labels["release.openshift.io/verify"]; ok {
+				return true
+			}
+		}
+		if tombstone, ok := obj.(cache.DeletedFinalStateUnknown); ok {
+			if prowJob, ok := tombstone.Obj.(*v1.ProwJob); ok {
+				if _, ok = prowJob.Labels["release.openshift.io/verify"]; ok {
+					return true
+				}
+			}
+		}
+		return false
+	}
+
+	prowJobInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: prowJobFilter,
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc: c.Enqueue,
+			UpdateFunc: func(old interface{}, new interface{}) {
+				oldProwJob, _ := old.(*v1.ProwJob)
+				newProwJob, _ := new.(*v1.ProwJob)
+
+				// Don't process resync or objects that are marked for deletion
+				if oldProwJob.ResourceVersion == newProwJob.ResourceVersion ||
+					!newProwJob.GetDeletionTimestamp().IsZero() {
+					return
+				}
+				c.Enqueue(new)
+			},
+			DeleteFunc: c.Enqueue,
+		},
+	})
+
+	return c
+}
+
+func (s *ProwJobController) SyncHandler(key string) error {
```

Comment:
Have your sync function take a key that corresponds to the thing that it is writing.  In this case, it is the ReleaseStatus.

This code then becomes
```
// for each prowjob in .status.something
//    if the prowjob state is terminal, do nothing
//    use the prowjoblister to get the prowjob in question
//    update the state.
```


This requires that a different controller write the list of prowjobs that have been created.  I think you have this information in two spots
1. the thing that creates the prowjob knows this
2. the thing that looks at a prowjob should be able to go from prowjob to the releasestatus